sphericart/src/sphericart_cuda.cpp

```diff
@@ -1,79 +1,108 @@
-#include <cuda.h>
-#include <cuda_runtime.h>
 #include <iostream>
 #include <stdexcept>
 
 #define _SPHERICART_INTERNAL_IMPLEMENTATION
+
+#include "dynamic_cuda.hpp"
+#include "sphericart.hpp"
 #include "cuda_base.hpp"
 #include "sphericart_cuda.hpp"
 
-/*host macro that checks for errors in CUDA calls, and prints the file + line
- * and error string if one occurs
- */
-
 using namespace sphericart::cuda;
 
-#define CUDA_CHECK(call)                                                                           \
-    do {                                                                                           \
-        cudaError_t cudaStatus = (call);                                                           \
-        if (cudaStatus != cudaSuccess) {                                                           \
-            std::cerr << "CUDA error at " << __FILE__ << ":" << __LINE__ << " - "                  \
-                      << cudaGetErrorString(cudaStatus) << std::endl;                              \
-            cudaDeviceReset();                                                                     \
-            exit(EXIT_FAILURE);                                                                    \
-        }                                                                                          \
-    } while (0)
+struct CudaError {
+    bool success;
+    std::string errorMessage;
+};
+
+/*
+    This code checks whether the cuda libraries we'd need to dynamically load are available on the host
+*/
+CudaError isCudaAvailable() {
+    void* cudaHandle = dlopen("libcuda.so", RTLD_NOW);
+    void* nvrtcHandle = dlopen("libnvrtc.so", RTLD_NOW);
+    void* cudartHandle = dlopen("libcudart.so", RTLD_NOW);
+
+    CudaError result;
+    result.success = true;
+
+    if (!cudaHandle) {
+        result.success = false;
+        result.errorMessage =
+            "Failed to load libcuda.so. Try running \"find /usr -name libcuda.so\" and "
+            "appending the directory to your $LD_LIBRARY_PATH environment variable.";
+    } else if (!cudartHandle) {
+        result.success = false;
+        result.errorMessage =
+            "Failed to load libcudart.so. Try running \"find /usr -name libcudart.so\" and "
+            "appending the directory to your $LD_LIBRARY_PATH environment variable.";
+    } else if (!nvrtcHandle) {
+        result.success = false;
+        result.errorMessage =
+            "Failed to load libnvrtc.so. Try running \"find /usr -name libnvrtc.so\" and "
+            "appending the directory to your $LD_LIBRARY_PATH environment variable.";
+    }
```

Comment:
I don't love that we duplicate the error messages from `dynamic_cuda.hpp`. Could we use the code from there instead? Maybe inside a `try {} catch {}` block

Reply:
the reason I've done it like this is because libcuda.so is in a different directory to the CUDA SDK.

Reply:
i've modified this now so that the class constructors return whether or not their respective handles are non-zero.